Upgrade Gulp 3 -> 4

xavdid opened the pull request:
Most of the commands aren't quite working yet, but this is the brunt of the work. Superseeds on #610, fixes #608. 

TODO: 

- [x] verify each public command's intended behavior
- [x] verify call tree is roughly the same (namely, that `clean` is always called when expected)
- [x] fix `File not found with singular glob` error
- [x] add default task
- [x] add docs

xavdid commented:
@carltongibson here's my WIP! I'm still getting a weird glob bug and I have some tidying to do (see todos in root), but this is most of the way there! Cleaned up some JS code while I was in there; hope that's ok.

Once it's ready for review, I'll go through and annotate it with notes about why certain changes were made.

xavdid commented:
Actually, I think I fixed the glob thing - now builds complete without error. 

Maybe a dumb question, but how do I actually see the result of the build? I see the generated `./static/build` folder, but don't see any html in there (only js, css, etc)

carltongibson commented:
Hey @xavdid Speedy work. Nice. Let me pull it and have a play. Thank you!

carltongibson commented:
Hey @xavdid — this looks to be along the right path. 

* Need to define a `default` task, so `gulp` by itself works. (This was `watch`) 
* Individual tasks need review, e.g. `styles` failed for me just testing... 

Good work! I'll try and add commits on-top of yours here and we can get this resolved. 

Thanks so much!

xavdid commented:
Ok, I've updated the code and fixed some bugs (that I had introduced). Build works consistently and I confirmed that the output of `gulp build` is identical to the existing version (via `diff`).

You mentioned you were seeing issues with tasks (such as `styles`) - it's running fine for me on my latest commit. Could you try it again and send over an error message if it's still a problem for you?

The other thing we could change here is which tasks are exposed at the top-level. Right now, everything is. But, is there ever a situation where someone would run a subtask (such as `styles:admin`, or even `styles`) by itself? It seems like the tasks are mostly called in `build` or `local`.

Anyway, if testing looks good, I can write up some docs and annotate the PR and we can get it out the door!

carltongibson commented:
@xavdid You're a hero 😀 let me have another play, and I'll follow up later. 🥇

carltongibson reviewed (commented):
Hi @xavdid — thanks again for this. Looks to be working well. 

* Node version 14.
* npm install

I think it’s only `watch`, `build` and `local` that are really used. 

If you could add some docs that you think help us and the next people along, that would be the 🍒 on top, but I think this is what we need. 

/cc @amakarudze 👍

xavdid commented on .node-version:
```diff
@@ -0,0 +1 @@
+14.16.1
```
The current LTS; Some of the deps are fairly old and might need another pass once v16 is active is LTS. You can probably wait until 14 is EOL though, which is a couple of years.

xavdid commented on README.md:
```diff
@@ -139,16 +139,47 @@ This means you shouldn't change any css files, but `.styl` files. They're in /st
 Autocompiling of `.styl` files to `.css`:
 
-    gulp watch
+    npx gulp watch
```
Best practice is not to install CLI tools globally; `npx` will use the local binary automatically (instead of having to type `./node_modules/.bin/gulp watch`).

xavdid commented on README.md:
````diff
@@ -139,16 +139,47 @@ This means you shouldn't change any css files, but `.styl` files. They're in /st
+The biggest gotcha is [async completion](https://gulpjs.com/docs/en/getting-started/async-completion#signal-task-completion). Each task much signal to gulp when it has finished. The easiest way to do this is using an `aysnc` function, **but** if your functionality uses gulp streams (most native gulp functionality does), then you should **not** use an `async` function. Instead, return the gulp str ...
+
+```js
+// WRONG - uses gulp's streams in an async function; subsequent tasks won't wait for completion correctly:
````
This was a big gotcha for me while working on this, so I figured I'd include it specifically.

None of this is important to the overall project README, so it could be moved into a `GULP-README.md` or something if you'd like.

xavdid commented on gulp/tasks/build.js:
```diff
@@ -1,4 +1,14 @@
+/**
+ * A production build is a local build with a different environment and revision control enabled
+ */
+const buildTask = series(setProduction, local, revisioning);
```
This is still the same steps, but re-uses what's already present in `local`

xavdid commented on gulp/tasks/clean.js:
```diff
@@ -1,9 +1,16 @@
-  clean = require('gulp-clean');
+"use strict";
 
+const del = require("del");
```
`gulp-clean` is deprecated, they suggested using this instead.

xavdid commented on gulp/tasks/clean.js:
```diff
@@ -1,9 +1,16 @@
+const cleanTask = async () => {
+  await del(config.paths.copy.dest.development);
+  await del(config.paths.copy.dest.production);
+  await del(config.paths.temp);
```
I added the development and temp paths here; figured those were worth cleaning too. Let me know if I should pull that out.

xavdid commented on gulp/tasks/environment.js:
```diff
@@ -1,12 +1,14 @@
-var development = environments.development;
-var production = environments.production;
+const { production, current } = require("gulp-environments");
+const log = require("fancy-log");
```
`gulp-util` is deprecated, they said use `fancy-log` instead.

xavdid commented on gulp/tasks/revisioning.js:
```diff
@@ -1,18 +1,27 @@
     .pipe(revCSS())
-    .pipe(gulp.dest(config.paths.build))
-    .pipe(revNapkin({verbose: false}))
+    .pipe(dest(config.paths.copy.dest.production))
```
explicitly use the production path. Changing this let me remove `config.paths.build`, since it was only used (and i clean, which now cleans everything anyway).

xavdid commented on gulp/config.js:
```diff
@@ -1,7 +1,6 @@
 module.exports = {
   paths: {
     project: './',
-    build: './static/build/',
```
duplicate of other key, see below

xavdid commented on gulp/tasks/styles.js:
```diff
@@ -1,34 +1,45 @@
+
+const config = require("../config");
+
+const stylusTask = () => {
```
no material changes here, just used a newer form of task declaration (namely, `task(...)` is deprecated)

xavdid commented on package.json:
```diff
@@ -1,5 +1,6 @@
 {
   "name": "djangogirls",
+  "private": true,
```
this prevents this from accidentially being published. There's nothing secret, but this isn't really made for publishing; it's just a collection of dependencies

xavdid commented on package.json:
```diff
@@ -10,21 +11,19 @@
+    "gulp-environments": "0.1.3",
     "gulp-rev": "^7.0.0",
-    "gulp-rev-css-url": "^0.1.0",
-    "gulp-rev-napkin": "^0.1.0",
```
Semver works funny with `^0.x.y` (you can have an in-range breaking version, see [this](https://docs.npmjs.com/cli/v6/using-npm/semver#caret-ranges-123-025-004)), so I pinned these to an exact version.

xavdid reviewed (commented):
Ok! Think we're all set here. I wrote docs for a future gulp developer and explained any other changes that  might not be self-evident. 

Let me know if you've got any questions or suggestions. We can also squash these into a single merge commit. I kept them separate for review readability, but I did some iterating that won't be necessary in the final historical record.

carltongibson reviewed (commented):
@xavdid — Thanks so much for the work here. This would have taken me an age to (not really) learn what I was doing, so 🎁 :)

All seems to work for me locally. @amakarudze Do you want to push it up when you have that cycle? 

Docs additions are handy for the future. 👍
